File: extension/src/lib/nvd.ts
// NVD CVE API 2.0 client with rate limiting and local applicability re-checking,
// mirroring the CLI's cve/nvd.go and cve/applies.go.

import { cpeParts, productFromCpeName, type Product } from "./cpe.ts";
import { Limiter, sleep } from "./limiter.ts";
import type { Vulnerability } from "./types.ts";
import { compareVersions, parseVersion } from "./version.ts";

export const DEFAULT_NVD_BASE_URL = "https://services.nvd.nist.gov/rest/json/cves/2.0";
export const NVD_DETAIL_URL = "https://nvd.nist.gov/vuln/detail/";

const PAGE_SIZE = 2000;
const WINDOW_MS = 30_000;
const PUBLIC_BUDGET = 5;
const KEYED_BUDGET = 50;
const MAX_ATTEMPTS = 3;
const THROTTLED_STATUSES = new Set([403, 429, 503]);

export interface NvdClientOptions {
  apiKey?: string;
  baseUrl?: string;
  fetchFn?: typeof fetch;
  retryDelayMs?: number;
}

interface NvdCpeMatch {
  vulnerable?: boolean;
  criteria?: string;
  versionStartIncluding?: string;
  versionStartExcluding?: string;
  versionEndIncluding?: string;
  versionEndExcluding?: string;
}

interface NvdMetric {
  baseSeverity?: string;
  cvssData?: { baseScore?: number; baseSeverity?: string };
}

interface NvdCve {
  id: string;
  published?: string;
  descriptions?: { lang: string; value: string }[];
  metrics?: {
    cvssMetricV40?: NvdMetric[];
    cvssMetricV31?: NvdMetric[];
    cvssMetricV30?: NvdMetric[];
    cvssMetricV2?: NvdMetric[];
  };
  references?: { url: string; tags?: string[] }[];
  configurations?: { nodes?: { cpeMatch?: NvdCpeMatch[] }[] }[];
}

interface NvdPage {
  totalResults?: number;
  vulnerabilities?: { cve: NvdCve }[];
}

class ThrottledError extends Error {}

export class NvdClient {
  private readonly apiKey: string;
  private readonly baseUrl: string;
  private readonly fetchFn: typeof fetch;
  private readonly retryDelayMs: number;
  private readonly limiter: Limiter;

  constructor(options: NvdClientOptions = {}) {
    this.apiKey = options.apiKey ?? "";
    this.baseUrl = options.baseUrl ?? DEFAULT_NVD_BASE_URL;
    // A bare `fetch` loses its receiver when stored as a property; bind it to the global.
    this.fetchFn = options.fetchFn ?? fetch.bind(globalThis);
    this.retryDelayMs = options.retryDelayMs ?? 6_000;
    this.limiter = new Limiter(this.apiKey ? KEYED_BUDGET : PUBLIC_BUDGET, WINDOW_MS);
  }

  /**
   * Returns every CVE that applies to the concrete version in cpeName. NVD pre-filters
   * by version server-side; each record is re-checked locally so unscoped wildcard
   * records do not flag every version.
   */
  async vulnerabilitiesFor(cpeName: string): Promise<Vulnerability[]> {
    const target = productFromCpeName(cpeName);
    const vulnerabilities: Vulnerability[] = [];
    let start = 0;
    for (;;) {
      const page = await this.fetchPage(cpeName, start);
      const items = page.vulnerabilities ?? [];
      for (const { cve } of items) {
        const affected = applicableRange(cve, target);
        if (affected !== null) {
          vulnerabilities.push({ ...toVulnerability(cve), affected_range: affected });
        }
      }
      start += items.length;
      if (items.length === 0 || start >= (page.totalResults ?? 0)) {
        return vulnerabilities;
      }
    }
  }

  private async fetchPage(cpeName: string, start: number): Promise<NvdPage> {
    const query = new URLSearchParams({
      cpeName,
      resultsPerPage: String(PAGE_SIZE),
      startIndex: String(start),
    });
    const headers: Record<string, string> = {};
    if (this.apiKey) {
      headers["apiKey"] = this.apiKey;
    }
    for (let attempt = 1; ; attempt++) {
      await this.limiter.wait();
      const response = await this.fetchFn(`${this.baseUrl}?${query}`, { headers });
      try {
        return await decodePage(response, cpeName);
      } catch (error) {
        if (!(error instanceof ThrottledError) || attempt >= MAX_ATTEMPTS) {
          throw error;
        }
        await sleep(this.retryDelayMs);
      }
    }
  }
}

async function decodePage(response: Response, cpeName: string): Promise<NvdPage> {
  if (response.ok) {
    return (await response.json()) as NvdPage;
  }
  if (THROTTLED_STATUSES.has(response.status)) {
    throw new ThrottledError(
      `NVD rate limit reached (HTTP ${response.status}) querying ${cpeName}; add an NVD API key in the extension options`,
    );
  }
  const reason = response.headers.get("message") ?? `HTTP ${response.status}`;
  throw new Error(`NVD rejected query for ${cpeName}: ${reason}`);
}

function toVulnerability(cve: NvdCve): Vulnerability {
  const vulnerability: Vulnerability = { id: cve.id, url: NVD_DETAIL_URL + cve.id };
  const [score, severity] = cvss(cve);
  if (score !== undefined) {
    vulnerability.score = score;
  }
  if (severity) {
    vulnerability.severity = severity;
  }
  if (cve.published) {
    vulnerability.published = cve.published;
  }
  const description = cve.descriptions?.find((d) => d.lang === "en")?.value;
  if (description) {
    vulnerability.description = description;
  }
  const advisory = advisoryUrl(cve);
  if (advisory) {
    vulnerability.advisory = advisory;
  }
  return vulnerability;
}

/** Score and severity from the newest CVSS version NVD has recorded. */
function cvss(cve: NvdCve): [number | undefined, string] {
  const metrics = cve.metrics ?? {};
  for (const list of [
    metrics.cvssMetricV40,
    metrics.cvssMetricV31,
    metrics.cvssMetricV30,
    metrics.cvssMetricV2,
  ]) {
    const metric = list?.[0];
    if (!metric) {
      continue;
    }
    const severity = metric.cvssData?.baseSeverity ?? metric.baseSeverity ?? "";
    return [metric.cvssData?.baseScore, severity.toUpperCase()];
  }
  return [undefined, ""];
}

/** The most authoritative reference: a vendor advisory, then a patch, then any. */
function advisoryUrl(cve: NvdCve): string {
  const references = cve.references ?? [];
  for (const tag of ["Vendor Advisory", "Patch", ""]) {
    const reference = references.find((r) => tag === "" || (r.tags ?? []).includes(tag));
    if (reference) {
      return reference.url;
    }
  }
  return "";
}

/** Scans every applicability statement for one that covers the product. */
export function applicableRange(cve: NvdCve, target: Product): string | null {
  for (const configuration of cve.configurations ?? []) {
    for (const node of configuration.nodes ?? []) {
      for (const match of node.cpeMatch ?? []) {
        const affected = matchApplies(match, target);
        if (affected !== null) {
          return affected;
        }
      }
    }
  }
  return null;
}

/**
 * A wildcard criteria with no version bounds is ignored: NVD carries many old records
 * whose applicability was never scoped, and they would flag every version ever released.
 */
export function matchApplies(match: NvdCpeMatch, target: Product): string | null {
  if (!match.vulnerable || !match.criteria) {
    return null;
  }
  const criteria = cpeParts(match.criteria);
  if (criteria.vendor !== target.vendor || criteria.product !== target.product) {
    return null;
  }
  if (criteria.version !== "*") {
    const exact = parseVersion(criteria.version);
    if (!exact || compareVersions(exact, target.version) !== 0) {
      return null;
    }
    return `== ${criteria.version}`;
  }
  return rangeApplies(match, target);
}

function rangeApplies(match: NvdCpeMatch, target: Product): string | null {
  const bounds: [string | undefined, string, (cmp: number) => boolean][] = [
    [match.versionStartIncluding, ">=", (c) => c >= 0],
    [match.versionStartExcluding, ">", (c) => c > 0],
    [match.versionEndIncluding, "<=", (c) => c <= 0],
    [match.versionEndExcluding, "<", (c) => c < 0],
  ];
  const parts: string[] = [];
  for (const [raw, op, ok] of bounds) {
    if (!raw) {
      continue;
    }
    const limit = parseVersion(raw);
    if (!limit || !ok(compareVersions(target.version, limit))) {
      return null;
    }
    parts.push(`${op} ${raw}`);
  }
  return parts.length === 0 ? null : parts.join(", ");
}

File: cve/nvd.go
package cve

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"slices"
	"strconv"
	"strings"
	"time"
)

// DefaultNVDBaseURL is the NVD CVE API 2.0 endpoint.
const DefaultNVDBaseURL = "https://services.nvd.nist.gov/rest/json/cves/2.0"

// NVDDetailURL prefixes a CVE ID to form its public NVD page.
const NVDDetailURL = "https://nvd.nist.gov/vuln/detail/"

const (
	userAgent       = "github.com/xZoroo/wappacvelyze/0.1"
	nvdPageSize     = 2000
	nvdWindow       = 30 * time.Second
	nvdPublicBudget = 5
	nvdKeyedBudget  = 50
	nvdRetryDelay   = 6 * time.Second
	nvdMaxAttempts  = 3
)

var errThrottled = errors.New("NVD rate limit reached")

// Vulnerability is a published CVE that applies to a detected technology version.
type Vulnerability struct {
	ID            string    `json:"id"`
	Severity      string    `json:"severity,omitempty"`
	Score         float64   `json:"score,omitempty"`
	Description   string    `json:"description,omitempty"`
	Published     string    `json:"published,omitempty"`
	URL           string    `json:"url"`
	Advisory      string    `json:"advisory,omitempty"`
	AffectedRange string    `json:"affected_range,omitempty"`
	KEV           *KEVEntry `json:"kev,omitempty"`
	// EPSS is FIRST's probability of exploitation within 30 days, when the database has it.
	EPSS           float64 `json:"epss,omitempty"`
	EPSSPercentile float64 `json:"epss_percentile,omitempty"`
	// Exploits names public exploit sources ("nuclei", "metasploit") that cover the CVE.
	Exploits []string `json:"exploits,omitempty"`
}

// NVDClient queries the NVD CVE API, pacing requests to stay within its rate limit.
type NVDClient struct {
	HTTP       *http.Client
	BaseURL    string
	APIKey     string
	limiter    *limiter
	retryDelay time.Duration
}

// NewNVDClient builds a client. Without an API key NVD allows 5 requests per 30 seconds;
// with one it allows 50.
func NewNVDClient(client *http.Client, apiKey string) *NVDClient {
	budget := nvdPublicBudget
	if apiKey != "" {
		budget = nvdKeyedBudget
	}
	return &NVDClient{
		HTTP:       client,
		BaseURL:    DefaultNVDBaseURL,
		APIKey:     apiKey,
		limiter:    newLimiter(budget, nvdWindow),
		retryDelay: nvdRetryDelay,
	}
}

// VulnerabilitiesFor returns every CVE that applies to the concrete version in cpeName.
// NVD pre-filters by version server-side; each result is re-checked locally so that
// unscoped wildcard records do not flag every version. The result is never nil: an empty
// slice means no known CVE applies.
func (c *NVDClient) VulnerabilitiesFor(
	ctx context.Context, cpeName string,
) ([]Vulnerability, error) {
	target, err := productFromCPEName(cpeName)
	if err != nil {
		return nil, err
	}
	vulns := []Vulnerability{}
	for start := 0; ; {
		page, err := c.fetchPage(ctx, cpeName, start)
		if err != nil {
			return nil, err
		}
		for _, item := range page.Vulnerabilities {
			affected, ok := item.CVE.applicableRange(target)
			if !ok {
				continue
			}
			v := item.CVE.toVulnerability()
			v.AffectedRange = affected
			vulns = append(vulns, v)
		}
		start += len(page.Vulnerabilities)
		if len(page.Vulnerabilities) == 0 || start >= page.TotalResults {
			return vulns, nil
		}
	}
}

func (c *NVDClient) fetchPage(ctx context.Context, cpeName string, start int) (*nvdPage, error) {
	req, err := c.newRequest(ctx, cpeName, start)
	if err != nil {
		return nil, err
	}
	for attempt := 1; ; attempt++ {
		if err := c.limiter.wait(ctx); err != nil {
			return nil, err
		}
		resp, err := c.HTTP.Do(req)
		if err != nil {
			return nil, fmt.Errorf("query NVD for %s: %w", cpeName, err)
		}
		page, err := decodePage(resp, cpeName)
		if !errors.Is(err, errThrottled) || attempt >= nvdMaxAttempts {
			return page, err
		}
		if err := sleep(ctx, c.retryDelay); err != nil {
			return nil, err
		}
	}
}

func (c *NVDClient) newRequest(ctx context.Context, cpeName string, start int) (*http.Request, error) {
	query := url.Values{}
	query.Set("cpeName", cpeName)
	query.Set("resultsPerPage", strconv.Itoa(nvdPageSize))
	query.Set("startIndex", strconv.Itoa(start))
	endpoint := c.BaseURL + "?" + query.Encode()
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint, nil)
	if err != nil {
		return nil, fmt.Errorf("build NVD request: %w", err)
	}
	req.Header.Set("User-Agent", userAgent)
	if c.APIKey != "" {
		req.Header.Set("apiKey", c.APIKey)
	}
	return req, nil
}

// decodePage parses an NVD response. NVD signals throttling with HTTP 403 and reports
// rejected queries in a "message" header.
func decodePage(resp *http.Response, cpeName string) (*nvdPage, error) {
	defer resp.Body.Close()
	switch resp.StatusCode {
	case http.StatusOK:
		var page nvdPage
		if err := json.NewDecoder(resp.Body).Decode(&page); err != nil {
			return nil, fmt.Errorf("decode NVD response for %s: %w", cpeName, err)
		}
		return &page, nil
	case http.StatusForbidden, http.StatusTooManyRequests, http.StatusServiceUnavailable:
		return nil, fmt.Errorf("%w (HTTP %d) querying %s; set NVD_API_KEY to raise the limit",
			errThrottled, resp.StatusCode, cpeName)
	default:
		reason := resp.Header.Get("message")
		if reason == "" {
			reason = resp.Status
		}
		return nil, fmt.Errorf("NVD rejected query for %s: %s", cpeName, reason)
	}
}

type nvdPage struct {
	TotalResults    int `json:"totalResults"`
	Vulnerabilities []struct {
		CVE nvdCVE `json:"cve"`
	} `json:"vulnerabilities"`
}

type nvdCVE struct {
	ID           string `json:"id"`
	Published    string `json:"published"`
	Descriptions []struct {
		Lang  string `json:"lang"`
		Value string `json:"value"`
	} `json:"descriptions"`
	Metrics struct {
		V40 []nvdMetric `json:"cvssMetricV40"`
		V31 []nvdMetric `json:"cvssMetricV31"`
		V30 []nvdMetric `json:"cvssMetricV30"`
		V2  []nvdMetric `json:"cvssMetricV2"`
	} `json:"metrics"`
	References []struct {
		URL  string   `json:"url"`
		Tags []string `json:"tags"`
	} `json:"references"`
	Configurations []struct {
		Nodes []struct {
			CPEMatch []nvdCPEMatch `json:"cpeMatch"`
		} `json:"nodes"`
	} `json:"configurations"`
}

// nvdMetric carries baseSeverity at the top level for CVSS v2 and inside cvssData for
// v3 and later.
type nvdMetric struct {
	BaseSeverity string `json:"baseSeverity"`
	CVSSData     struct {
		BaseScore    float64 `json:"baseScore"`
		BaseSeverity string  `json:"baseSeverity"`
	} `json:"cvssData"`
}

func (c nvdCVE) toVulnerability() Vulnerability {
	v := Vulnerability{
		ID:        c.ID,
		Published: c.Published,
		URL:       NVDDetailURL + c.ID,
		Advisory:  c.advisory(),
	}
	v.Score, v.Severity = c.cvss()
	for _, d := range c.Descriptions {
		if d.Lang == "en" {
			v.Description = d.Value
			break
		}
	}
	return v
}

// cvss returns the score from the newest CVSS version NVD has recorded.
func (c nvdCVE) cvss() (float64, string) {
	for _, metrics := range [][]nvdMetric{c.Metrics.V40, c.Metrics.V31, c.Metrics.V30, c.Metrics.V2} {
		if len(metrics) == 0 {
			continue
		}
		m := metrics[0]
		severity := m.CVSSData.BaseSeverity
		if severity == "" {
			severity = m.BaseSeverity
		}
		return m.CVSSData.BaseScore, strings.ToUpper(severity)
	}
	return 0, ""
}

// advisory picks the most authoritative reference: a vendor advisory, then a patch, then
// whatever NVD lists first.
func (c nvdCVE) advisory() string {
	for _, tag := range []string{"Vendor Advisory", "Patch", ""} {
		for _, ref := range c.References {
			if tag == "" || slices.Contains(ref.Tags, tag) {
				return ref.URL
			}
		}
	}
	return ""
}
